devicetypes/smartthings/cooper-receptacle.src/cooper-receptacle.groovy

```diff
@@ -0,0 +1,172 @@
+
+		input ("protection", "enum", title: "Child Lockout", description: "Disabled",
+				required: false, options:["disabled": "Disabled", "sequence": "Protection by Sequence", "noOperation": "No local operation"])
+	}
+
+	tiles(scale: 2) {
+		multiAttributeTile(name:"switch", type: "lighting", width: 6, height: 4, canChangeIcon: false){
+			tileAttribute ("device.switch", key: "PRIMARY_CONTROL") {
+				attributeState ("on", label: '${name}', action: "switch.off",
+						icon: "st.switches.switch.on", backgroundColor: "#00A0DC")
+				attributeState ("off", label: '${name}', action: "switch.on",
+						icon: "st.switches.switch.off", backgroundColor: "#ffffff")
+			}
+		}
+		standardTile("refresh", "device.switch", width: 2, height: 2, inactiveLabel: false, decoration: "flat") {
+			state "default", label:'', action:"refresh.refresh", icon:"st.secondary.refresh"
+		}
+
+		main "switch"
+		details(["switch","refresh"])
+	}
+}
+
+def updated(){
+	def cmds = []
+	cmds << hubAction(zwave.configurationV1.configurationSet(configurationValue: [getPowerUpStateValue()],
+			parameterNumber: 5, size :1).format())
+	cmds << hubAction(zwave.protectionV1.protectionSet(protectionState: getProtection()).format())
+	cmds << hubAction(zwave.manufacturerSpecificV2.manufacturerSpecificGet().format())
+	sendHubCommand(cmds, 300)
+}
+
+def on() {
+	def cmd = []
+	cmd << hubAction(zwave.basicV1.basicSet(value: 0xFF).format())
+	cmd << hubAction(zwave.basicV1.basicGet().format())
+	sendHubCommand(cmd, 300)
+}
+
+def off() {
+	def cmd = []
+	cmd << hubAction(zwave.basicV1.basicSet(value: 0x00).format())
+	cmd << hubAction(zwave.basicV1.basicGet().format())
+	sendHubCommand(cmd, 300)
+}
+
+def refresh() {
+	def cmd = []
+	cmd << hubAction(zwave.basicV1.basicGet().format())
+	cmd << hubAction(zwave.configurationV1.configurationGet(parameterNumber: 5).format())
+	cmd << hubAction(zwave.protectionV1.protectionGet().format())
+	sendHubCommand(cmd, 300)
+}
+
+def parse(String description) {
+	def cmd = zwave.parse(description, [0x75:1])
+	if (cmd) {
+		zwaveEvent(cmd)
+	}
+	return []
+}
+
+def zwaveEvent(physicalgraph.zwave.commands.basicv1.BasicReport cmd) {
+	sendEvent(name: "switch", value: cmd.value ? "on" : "off")
+}
+
+def zwaveEvent(physicalgraph.zwave.commands.basicv1.BasicSet cmd) {
+	sendEvent(name: "switch", value: cmd.value ? "on" : "off")
+}
+
+def zwaveEvent(physicalgraph.zwave.commands.manufacturerspecificv2.ManufacturerSpecificReport cmd) {
+	if (cmd.manufacturerName) {
+		updateDataValue("manufacturer", cmd.manufacturerName)
+	}
+	if (cmd.productTypeId) {
+		updateDataValue("productTypeId", cmd.productTypeId.toString())
+	}
+	// productId 0 is a valid value
+	if (cmd.productId != null) {
+		updateDataValue("productId", cmd.productId.toString())
+	}
+}
+
+def zwaveEvent(physicalgraph.zwave.commands.configurationv2.ConfigurationReport cmd) {
+	//settings are read-only, so send event if settings are out of sync
```

Comment:
For here and below, you can do

`device.updateSetting("powerUpState", cmd.configurationValue[0])`

It's not documented, but I've tested that it works.

Reply:
Thank You very much. That's huge help.

Reply:
I tried: 

`device.updateSetting("delayedOffTime", cmd.configurationValue[0])`

This results in:

>  c24c73e7-270b-4629-a1ed-3fc5818128d4 2:12:01 PM: error org.hibernate.LazyInitializationException: failed to lazily initialize a collection of role: physicalgraph.device.Device.preferences, no session or session was closed 

What am I doing wrong?

Reply:
I guess this is why it's undocumented. 

As an alternate solution, try doing what is done here: https://github.com/SmartThingsCommunity/SmartThingsPublic/blob/master/devicetypes/smartthings/ge-link-bulb.src/ge-link-bulb.groovy

Specifically look at the relationship between `dimRate` and `state.rate`.

Essentially, the DTH has a duplication of the preference in `state.____`. Updates from the user change this value as well as updates from the device, and it is treated as the canonical value for logic purposes.

Reply:
Does zigbee handle state.rate updates from device internally? This DTH seems to handle only commands with resultMap.name other than "level".

We can use mechanism you describe like this:
- Add value tile for each property, so that user can go to preferences and change them if he does not like those values.
- Hold received configuration value in state and update preferences in next refresh call (or any other capability method if it works there; we would have to check if it works when called by SmartApp too)
- Combination of the above, which could be confusing as values in preferences will or will not match those in value tile depending on other user and/or SmartApp actions.

Reply:
I think option 2 is the best, but I would still use the state value as the authoritative value instead of the preference. So in that case the three main points are:

* `update` gets called, so set the state value and send the zwave set.

* zwave report received so update the state value

* capability method called so set the preference to the state value if they're different (`updatePreference` does seem to work in any capability method)

Reply:
For now, I added attributes and standard tiles to display current configuration.

I tested option 2. I added SmartLighting automation that turned the receptacle on/off when remote switch is turned on/off. Unfortunately this resulted in org.hibernate.LazyInitializationException each time on() or off() method was called by SmartLighting SmartApp (by switching remote switch on/off).